os/drivers/spi/spi_uio.c

```diff
@@ -0,0 +1,233 @@
+
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <tinyara/config.h>
+#include <assert.h>
+#include <errno.h>
+#include <debug.h>
+#include <tinyara/fs/fs.h>
+#include <tinyara/fs/ioctl.h>
+#include <tinyara/spi/spi.h>
+
+#include <semaphore.h>
+
+#define SPI_ENTRY printf("-->%s\t%s:%d\n", __FUNCTION__, __FILE__, __LINE__)
+/******************************************************************************
+ * Private Variables
+ *****************************************************************************/
+static int spi_uio_open(FAR struct file *filep);
+static int spi_uio_close(FAR struct file *filep);
+static int spi_uioctrl(FAR struct file *filep, int cmd, unsigned long arg);
+static ssize_t spi_uioread(FAR struct file *filep, FAR char *buffer, size_t buflen);
+static ssize_t spi_uiowrite(FAR struct file *filep, FAR const char *buffer, size_t buflen);
+
+static const struct file_operations g_spiops = {
+	spi_uio_open,				/* open */
+	spi_uio_close,				/* close */
+	spi_uioread,				/* read */
+	spi_uiowrite,				/* write */
+	NULL,						/* seek */
+	spi_uioctrl,				/* ioctl */
+#ifndef CONFIG_DISABLE_POLL
+	NULL						/* poll */
+#endif
+};
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+
+/****************************************************************************
+ * Name: spidrvr_open
+ ****************************************************************************/
+
+static int spi_uio_open(FAR struct file *filep)
+{
+    SPI_ENTRY;
+	FAR struct inode *inode;
+	FAR struct spi_dev_s *dev;
+
+	/* Get our private data structure */
+	DEBUGASSERT(filep != NULL && filep->f_inode != NULL);
+	inode = filep->f_inode;
+
+	dev = (FAR struct spi_dev_s *)inode->i_private;
+	DEBUGASSERT(dev);
+
+	return 0;
+}
+
+/****************************************************************************
+ * Name: spidrvr_close
+ ****************************************************************************/
+
+static int spi_uio_close(FAR struct file *filep)
+{
+	SPI_ENTRY;
+	FAR struct inode *inode;
+	FAR struct spi_dev_s *dev;
+
+	/* Get our private data structure */
+
+	DEBUGASSERT(filep != NULL && filep->f_inode != NULL);
+	inode = filep->f_inode;
+
+	dev = (FAR struct spi_dev_s *)inode->i_private;
+	DEBUGASSERT(dev);
+
+	return 0;
+}
+
+
+/****************************************************************************
+ * Name: spi_uioread
+ *
+ * Description:
+ *   Read SPI Data. This function is automatically called when you're doing
+ *   read() API on userspace.
+ *
+ ****************************************************************************/
+
+static ssize_t spi_uioread(FAR struct file *filep, FAR char *buffer, size_t buflen)
```

Comment:
semaphore should be adde in read & write function to prevent functionality error

Reply:
Unfortunately there are many driver codes that are not protected by sem(ex. i2c...). so I suggest that fix all these vulnerabilities next PR. furthermore I'll adapt spi from the latest nuttx code later.